pkg/sql/sem/plpgsqltree/plpg_visitor.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package plpgsqltree
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/server/telemetry"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqltelemetry"
+	"github.com/cockroachdb/errors"
+)
+
+// PLpgSQLStmtCounter is used to accurately report telemetry for plpgsql
+// statements. We can not use the counters due to counters needing to
+// be reset after every statement using reporter.ReportDiagnostics.
+type PLpgSQLStmtCounter map[string]int
+
+func (p *PLpgSQLStmtCounter) String() string {
+	var buf strings.Builder
+	counter := *p
+
+	// Sort the counters to avoid flakes in test
+	keys := make([]string, 0)
+	for k := range counter {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		buf.WriteString(fmt.Sprintf("%s: %d\n", k, counter[k]))
+
+	}
+	return buf.String()
+}
+
+type plpgsqlVisitor struct {
+	ctx  context.Context
+	stmt PLpgSQLStatement
+	// StmtCnt captures telemetry for plpgsql statements that can be returned
+	// during parser test.
+	StmtCnt PLpgSQLStmtCounter
+	// Err captures errors while stmts are being walked.
+	Err error
+}
+
+func MakePLpgSQLVisitor() plpgsqlVisitor {
+	return plpgsqlVisitor{ctx: context.Background(), StmtCnt: PLpgSQLStmtCounter{}}
+}
+
+type walkableStmt interface {
+	PLpgSQLStatement
+	walkStmt(plpgsqlVisitor) PLpgSQLStatement
+}
+
+var _ walkableStmt = &PLpgSQLStmtIf{}
+
+// WalkStmt walks plpgsql statements to ensure that telemetry is captured for nested statements.
+func WalkStmt(v plpgsqlVisitor, stmt PLpgSQLStatement) (newStmt PLpgSQLStatement, changed bool) {
+	wStmt, isWalkable := stmt.(walkableStmt)
+
+	if !isWalkable {
```

Comment:
Even though I think it should work, I don't think this is quite right... It feels weird that you do the counting explicitly here while also have the visitor do the counting. But in general I think you're moving toward the right direction.

Ideally, there should be a `StmtVisitor` interface which has `Visit(stmt walkableStmt)` method which will be implemented to have the counting logic. Then, in this `WalkStmt` function, you just call `stmt.walkStmt(visitor)`, and in `stmt.walkStmt(visitor` you do `visitor.Visit(stmt)`. Then call `WalkStmt` on each sub statements if there is any.

I know that in the SQL world the visitor actually only visit expressions, while the statement walking walks on the statements and sub-clauses. For telemetry counting purpose, our visitor can be simple and just visit the statement without considering the expressions.